### ISSUE TITLE
GIMP color dialog keeps appearing in the first work space.

### ISSUE THREAD

**Issue by crocket** [xmonad/xmonad#107]
### Problem Description

My configuration sends every window with appName starting with `gimp` to `DRAW` workspace.
`DRAW` workspace is the second workspace that lives on my second monitor.
`MAIN` workspace is the first workspae on my first monitor.

Other gimp dialogs appear on `DRAW` workspace.
But, a dialog window with the attribute `WM_WINDOW_ROLE(STRING) = "gimp-toolbox-color-dialog"` keeps appearing in the first workspace after appearing in `DRAW` workspace for the first time it appears.

Gimp 2.9.4
xmonad 0.13
xmonad-contrib 0.13

You can create a color dialog from the toolbox by clicking one of two overlapping color rectangles in the toolbox shown below.

![gimp-color-dialog](https://cloud.githubusercontent.com/assets/748856/25652159/bd50261e-3021-11e7-95ae-38f3353c0353.png)

### Partial Configuration File

```haskell
{-# LANGUAGE DuplicateRecordFields #-}
import XMonad
import qualified XMonad.StackSet as W
-- xmonad-contrib
import XMonad.Util.Run (spawnPipe)
import XMonad.Hooks.ManageDocks (docks, avoidStruts)
import XMonad.Hooks.DynamicLog
import XMonad.Hooks.EwmhDesktops (ewmh, fullscreenEventHook)
import XMonad.Layout.NoFrillsDecoration
import XMonad.Layout.Spacing
import XMonad.Layout.Tabbed (tabbed)
import XMonad.Prompt.ConfirmPrompt
import XMonad.Prompt.Workspace
import XMonad.Util.EZConfig (additionalKeysP)
import XMonad.Layout.Renamed (renamed, Rename(Replace, CutWordsLeft))
import XMonad.Actions.PerWorkspaceKeys (chooseAction)
import qualified XMonad.Util.NamedScratchpad as N
import qualified XMonad.Hooks.ManageHelpers as MH
-- my lib
import qualified XMonad.MyTheme as T
import qualified XMonad.Maim as Maim
import qualified XMonad.Watson as Watson
import qualified XMonad.Dmenu as D
import qualified XMonad.Gxmessage as G
import qualified XMonad.Timer as Timer
import XMonad.Scratchpads
import qualified XMonad.Option as Opt
import XMonad.Tmux (TmuxWindow(..))
-- GHC
import System.IO (hPutStrLn)
import System.Exit (exitWith, ExitCode(..))
import qualified Data.Map.Strict as M
import Data.List (isPrefixOf)

myModMask :: KeyMask
myModMask = mod4Mask

main :: IO ()
main = do
  xmproc <- spawnPipe "xmobar"
  xmonad $ ewmh $ docks def
    { terminal = "st tmux new-session -A -D -s main"
    , modMask = myModMask
    , borderWidth = 0
    , workspaces = ["MAIN", "DRAW", "LANG"]
    , layoutHook = avoidStruts $ addTopBar $
      (renamed [Replace "Tall"] $ smartSpacing 7 $ Tall 1 (3/100) (1/2)) |||
      (renamed [Replace "Tabbed"] $ tabbed shrinkText T.tabs)
    , logHook = dynamicLogWithPP . N.namedScratchpadFilterOutWorkspacePP $
                xmobarPP
                { ppOutput = hPutStrLn xmproc
                , ppTitle = xmobarColor "green" "" . shorten 50
                }
    , focusFollowsMouse = True
    , clickJustFocuses = False
    , handleEventHook = fullscreenEventHook <+> handleEventHook def
    , manageHook =  myManageHook <+> N.namedScratchpadManageHook scratchpads
    , startupHook = spawn "compton"
    }

-- Windowing Policy
myManageHook = composeAll
  [ appName =? "xmessage" --> MH.doCenterFloat
  , MH.isDialog --> MH.doCenterFloat
  , className =? "mpv" --> doFloat
  , appName =? "gxmessage" --> doFloat
  , appName =? "ibus-setup" --> MH.doCenterFloat
  -- gimp
  , isPrefixOf "gimp-toolbox" <$> role --> doFloat
  , isPrefixOf "gimp-dock" <$> role --> doFloat
  , isPrefixOf "gimp" <$> appName --> doShift "DRAW"
  ]
  where role = stringProperty "WM_WINDOW_ROLE"

-- Layout elements
addTopBar = renamed [CutWordsLeft 1] . noFrillsDeco shrinkText T.topBar
```

**Comment by crocket** [comment]
Inserting `role =? "gimp-toolbox-color-dialog" --> doShift "DRAW"` into `myManageHook` doesn't solve the issue.

Many gimp dialogs including preferences dialog ignore `MH.isDialog --> MH.doCenterFloat`. Not all gimp dialogs ignore this.

In `Edit -> Preferences -> Interface -> Window Management`, I see `Open windows on the same monitor they were open before`. This option is not checked. When it's not checked, all new gimp windows should appear on the current monitor that gimp occupies. GIMP fails to catch its current monitor and just pushes its color dialog to the first monitor.

At least for color dialog and preferences dialog, the positions are controlled by gimp directly.

Is there a way to make xmonad override gimp's window management?

**Comment by geekosaur** [comment]
You might try this without `compton` running; I have observed it getting confused and pushing stuff onto the wrong monitor, especially when the window already exists (the fact that it does not happen the first time the window appears suggests to me that `gimp` is keeping it in `WITHDRAWN` state after it is first opened).

If this is happening, there's nothing `xmonad` can do about it; `compton` as compositor is the only program that actually renders windows to the display.

**Comment by crocket** [comment]
I disabled and killed compton, but the issue still persists.

However, without compton, I can see that the color dialog first appears in the current workspace(DRAW) for a very short time and is moved to the first monitor.

**Comment by geekosaur** [comment]
That's likely to be the application's fault then, possibly because our workspace model isn't the same as Gnome's (and the pervasive "we are the only ones who matter" of anyone related to gnome/gtk any more; don't even bother reporting this as a bug in gimp, as far as they are concerned the "bug" is you used it with anything but Holy Gnome).

There are ways you could try to force the issue. The most likely result is you end up with xmonad and gimp having a focus war.

**Comment by crocket** [comment]
What about sleeping for 0.3 second in another thread and shifting gimp dialogs to a certain workspace?
I would see some gimp dialogs shift to different workspaces twice, but it'd be better than manually moving windows with a mouse cursor.

I am also thinking about asking gimp developers to make an option to disable position management for dialogs.

**Comment by crocket** [comment]
I found a way to work around the issue.

In `Edit -> Preferences -> Interface -> Window Management`,

* Click `Reset Saved Window Positions to Default Values`
* Uncheck `Saved window positions on exit`
* Restart `gimp`
* color dialog starts appearing in the correct workspace and monitor.

If windows positions aren't saved on exit, I have to manually save dock window arrangements via `Save Window Positions Now` in `Edit -> Preferences -> Interface -> Window Management` without opening color dialog and restart gimp. If `Save Window Positions Now` is clicked after color dialog is opened even once and gimp is restarted, the color dialog keeps appearing in the first monitor, again. Thus, I have to click `Reset Saved Window Positions to Default Values` again.

It seems gimp window management manages both dialog window positions and toolbox/dock window arrangements.

While dialog positions are not saved, `MH.doCenterFloat` and `doShift` are respected by gimp dialogs.

I reported this issue on https://bugzilla.gnome.org/show_bug.cgi?id=782194
